pkg/rke2/kp.go

```diff
@@ -0,0 +1,52 @@
+package rke2
+
+import (
+	"context"
+
+	"github.com/rancher/k3s/pkg/cli/cmds"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+
+	"github.com/k3s-io/helm-controller/pkg/generated/controllers/helm.cattle.io"
+
+	"k8s.io/client-go/tools/clientcmd"
+
+	"github.com/sirupsen/logrus"
+	"github.com/urfave/cli"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const kubeProxyChart = "rke2-kube-proxy"
+
+// setKubeProxyDisabled determines if a cluster already has kube proxy deployed as a chart, if so
+// disables running kubeproxy as a static pod.
+func setKubeProxyDisabled(clx *cli.Context, cfg *cmds.Server) func(context.Context, <-chan struct{}, string) error {
+	return func(ctx context.Context, apiServerReady <-chan struct{}, kubeConfigAdmin string) error {
+		go func() {
+			<-apiServerReady
+			logrus.Info("Checking for kube proxy as a chart")
+
+			restConfig, err := clientcmd.BuildConfigFromFlags("", kubeConfigAdmin)
+			if err != nil {
+				logrus.Fatalf("kp: new k8s cient: %s", err.Error())
+			}
+
+			hc, err := helm.NewFactoryFromConfig(restConfig)
+			if err != nil {
+				logrus.Fatalf("kp: new helm client: %s", err.Error())
+			}
+			if _, err := hc.Helm().V1().HelmChartConfig().Get(metav1.NamespaceSystem, kubeProxyChart, metav1.GetOptions{}); err != nil {
+				if apierrors.IsNotFound(err) {
+					logrus.Infof("%[1]s HelmChartConfig not found, disabling %[1]s", kubeProxyChart)
+					clx.Set("disable", kubeProxyChart)
+					return
+				}
+				logrus.WithError(err).Fatalf("kp: failed to check for %s HelmChartConfig", kubeProxyChart)
+				return
+			}
+			logrus.Infof("%s HelmChartConfig found, disabling embedded kube-proxy", kubeProxyChart)
+			cfg.DisableKubeProxy = true
```

Comment:
I am slightly concerned that kube-proxy might already be queued up to be started by the time we get here, but this should be easy enough to test.

We might also need to add some logic to remove the kube-proxy static chart if it was enabled and has been disabled on this startup. The static pod executor currently lacks this sort of logic, so there are bits of code scattered around to clean up manifests when etcd or the apiserver are disabled. This can probably be fixed later, and more holistically.

Reply:
@brandond how do we want to proceed?

Reply:
I would probably just verify that this is actually working - install RKE2 1.21 from tarball, add a HelmChartConfig for rke2-kube-proxy, then install a CI build and verify that it disables the kube-proxy static chart as desired.

Reply:
@brandond I tested this and verified that the kube-proxy static pod doesn't start up.